docs/data/data-grid/server-side-data/tree-data.md

```diff
@@ -2,14 +2,67 @@
+# Data Grid - Server-side tree data [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan 'Pro plan')
 
 <p class="description">Tree data lazy-loading with server-side data source.</p>
 
-:::warning
-This feature isn't implemented yet. It's coming.
+To dynamically load tree data from the server, including lazy-loading of children, you must create a data source and pass the `unstable_dataSource` prop to the Data Grid, as detailed in the [overview section](/x/react-data-grid/server-side-data/).
 
-👍 Upvote [issue #3377](https://github.com/mui/mui-x/issues/3377) if you want to see it land faster.
+The data source also requires some additional props to handle tree data, namely `getGroupKey` and `getChildrenCount`. If the children count is not available for some reason, but there are some children, `getChildrenCount` should return `-1`.
 
-Don't hesitate to leave a comment on the same issue to influence what gets built. Especially if you already have a use case for this component, or if you are facing a pain point with the [currently proposed workaround](https://mui.com/x/react-data-grid/tree-data/#children-lazy-loading).
+```tsx
+const customDataSource: GridDataSource = {
+  getRows: async (params) => {
+    // Fetch the data from the server
+  },
+  getGroupKey: (row) => {
+    // Return the group key for the row, e.g. `name`
+    return row.name;
+  },
+  getChildrenCount: (row) => {
+    // Return the number of children for the row
+    return row.childrenCount;
+  },
+};
+```
+
+Following tree-data example supports filtering, sorting, and pagination on the server. It also caches the data by default.
+
+{{"demo": "ServerSideTreeData.js", "bg": "inline"}}
+
+:::info
+The data source demos use a utility function `useMockServer` to simulate the server-side data fetching. In a real-world scenario, you should replace this with your own server-side data fetching logic.
+
+Open info section of the browser console to see the requests being made and the data being fetched in response.
 :::
+
+## Error handling
+
+For each row group expansion, the data source is called to fetch the children. If an error occurs during the fetch, the grid will display an error message in the row group cell. `unstable_onDataSourceError` is also triggered with the error and the fetch params.
+
+The demo below shows a toast apart from the default error message in the grouping cell. Cache has been disabled in this demo for simplicity.
+
+{{"demo": "ServerSideTreeDataErrorHandling.js", "bg": "inline"}}
+
+## Group expansion
+
+The idea behind the group expansion is the same as explained in the [Row grouping](/x/react-data-grid/row-grouping/#group-expansion) section. The difference is that the data is not initially available and is fetched automatically after the Data Grid is mounted based on the props `defaultGroupingExpansionDepth` and `isGroupExpandedByDefault` in a waterfall manner.
```

Comment:
That's a good idea in general, for this specific case, we [use](https://mui.com/x/react-data-grid/tree-data/#group-expansion) the same referencing in the client-side tree data as well.

Since the concepts and linked pages are relevant to each other, I feel that referencing them might be a win-win as it'll allow to (1) reduce the duplication and (2) improve the users' knowledge about various Data Grid linked features.

What do you think?